builder/builder.go

```diff
@@ -643,6 +656,43 @@ func (b *Builder) AddBundles(bundles []string, alllocal bool, allupstream bool,
 	return nil
 }
 
+// ListBundles prints out the bundles in the mix, all bundles in the mix
+// including recursive includes, all bundles available in local bundles, or
+// all bundles available in upstream bundles
+func (b *Builder) ListBundles(full bool, local bool, upstream bool) error {
+	var bundles []string
+	var err error
+
+	switch {
+	case local:
+		bundles, err = b.getDirBundleList(b.Lbundledir)
+		if err != nil {
+			return errors.New("Error retreiving local bundle list")
+		}
+	case upstream:
+		bundles, err = b.getDirBundleList(b.getUpstreamBundlesPath(b.Clearver))
+		if err != nil {
+			return errors.New("Error retreiving local bundle list")
+		}
+	case full:
+		bundles, err = b.getMixBundlesListFull()
+		if err != nil {
+			return errors.New("Error retreiving mix bundle list")
+		}
+	default:
```

Comment:
This should not be the default. The default is "I want to know what bundles I'm about to spend X amount of time building," which should include every single bundle that is going to get built (includes, deps, etc). Unless that is what this is already doing and I misunderstood the code :)

Reply:
Nah, you understood correctly. Valid point.

I fall on the other side of the fence. There has _never been_ a Mix Bundle List before, so I know your mindset is definitely firmly in "what bundles are going in my mix" and not "what bundles do I _want_ in my mix". And it's that second thought that I'm trying to push things toward, from a UX perspective. That's large part of the motivation for the Mix Bundle List -- stating only the bundles you care about, and letting the tool figure out what dependencies it needs to pull in to meet that.  It's _that_ list that all of the `mixer bundle add/delete/edit/whatever` commands are intended to deal with.

For debugging or validation purposes, it's invaluable to be able to see (without trudging through leftover build artifacts) what all bundles will be fully in the list, but IMHO that's not what users should _normally_ be thinking about.

When someone runs `mixer bundle add editors`, for example, it's _only_ the `editors` bundle that is added to their Mix Bundle List. So if the user then does `mixer bundle list`, having it also spit back out `python-basic` and `python3-basic` feels really confusing. If they then were to go "Oh weird, I don't want python! `mixer bundle remove python-basic`, **nothing would happen**, because python-basic was never in the list, only `editors`.

Reply:
I disagree, you are about *all* your inputs, not just what you think you added. There have been multiple emails and issues opened where people try to add a single bundle (sysadmin-basic, or editors) which pulls in python, python3, etc, which are massive, and to their dismay their image/container is now GB in size vs just a few hundred MB as they thought. It's been a pain point for us for a while. As someone who is building an OS I want to know what things are going into my build and why. 
Yes, before your update to recurse includes and bring them in, it would crash and say "you're missing these bundles too," and you would have to add them manually. However, you're still aware of exactly what's being added and why, and finding out what your next build will have means typing  "ls mix-bundles". 
I don't see a reason to hide this from people at all, it creates ambiguity and more commands to run. The ideal  scenario IMO is a local-bundles folder, with users' custom bundle defs, the clr-bundles-VER folder with upstream bundles, and the mixbundles.list which specifies ALL bundles going into a given build.  The list can be more detailed specifying which bundles come from where etc etc, but at the least it gives an easily query-able data point for the builder input. Or you generate the mix-bundles directory each time the mixbundles.list is updated and data can be read from there. There has to be some persistent data of what will actually  be built, and if a user then wants to know "which ones only came from me, or only from upstream" they can pass the appropriate flags then.